torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -148,29 +148,47 @@ const std::exception& ProcessGroupNCCL::WorkNCCL::exception() const {
       "isCompleted() and wait() will either succeed or throw");
 }
 
+std::unordered_set<ssize_t> ProcessGroupNCCL::processGroups_;
+ssize_t ProcessGroupNCCL::processGroupCounter_ = -1;
+
 ProcessGroupNCCL::ProcessGroupNCCL(
     const std::shared_ptr<Store>& store,
     int rank,
     int size)
     : ProcessGroup(rank, size), store_(store) {
-  C10D_CUDA_CHECK(cudaGetDeviceCount(&numGPUs_));
-  thcState_ = ::at::globalContext().lazyInitCUDA();
+  // C10D_CUDA_CHECK(cudaGetDeviceCount(&numGPUs_));
+  thcState_ = ::at::globalContext().getTHCState();
+  if (!thcState_) {
+    thcState_ = ::at::globalContext().lazyInitCUDA();
+  }
+  // Generate the Process Group ID for current PG, this needs to be identical
+  // for all processes
+  std::unique_lock<std::mutex>(pgTrackingLock_);
+  while (true) {
+    ++processGroupCounter_;
+    if (processGroups_.count(processGroupCounter_) == 0) {
+      processGroups_.insert(processGroupCounter_);
+      processGroupId_ = std::to_string(processGroupCounter_);
+      return;
+    }
+  }
 }
 
-ProcessGroupNCCL::~ProcessGroupNCCL() {}
+ProcessGroupNCCL::~ProcessGroupNCCL() {
+  std::unique_lock<std::mutex>(pgTrackingLock_);
+  processGroups_.erase(std::stoull(processGroupId_));
+}
 
-void ProcessGroupNCCL::broadcastUniqueNCCLId(
-    const std::string& devicesKey,
-    ncclUniqueId* ncclId) {
+void ProcessGroupNCCL::broadcastUniqueNCCLId(ncclUniqueId* ncclId) {
   // Rank 0 writes to the store as bcast
   if (rank_ == 0) {
     auto ncclIdVal = std::vector<uint8_t>(
         reinterpret_cast<uint8_t*>(ncclId),
         reinterpret_cast<uint8_t*>(ncclId) + NCCL_UNIQUE_ID_BYTES);
-    store_->set(devicesKey, ncclIdVal);
+    store_->set(processGroupId_, ncclIdVal);
```

Comment:
The process group ID is unique per process group. If this function gets called more than once, the processes with rank != 0 will see the old `ncclIdVal`. If we have one NCCL communicator per process group per device list, we have to scope the keys in the store as such, to prevent races. The key should consist of both the process group ID and the devices key.

Reply:
The device list won't help here, since they can differ from process to process. I guess we shouldn't cache the NCCL ID in the store anyway since they can only be used one time